addons/package_hierarchy/models/package_links.py

```diff
@@ -0,0 +1,182 @@
+from odoo.exceptions import ValidationError
+
+
+class PackageHierarchyLink(models.Model):
+    _name = "package.hierarchy.link"
+    _description = "Links for constructing package hierarchies"
+    _order = "id"
+
+    _parent_name = "parent_id"
+
+    name = fields.Char(index=True, store=True, compute="_compute_name")
+
+    parent_id = fields.Many2one(
+        "stock.quant.package", string="Parent package", ondelete="cascade", check_company=True
+    )
+    child_id = fields.Many2one(
+        "stock.quant.package", string="Child package", ondelete="cascade", check_company=True
+    )
+    move_line_ids = fields.Many2many(
+        "stock.move.line",
+        column1="link_id",
+        column2="move_line_id",
+        string="Stock move line",
+        ondelete="cascade",
+        check_company=True,
+    )
+    company_id = fields.Many2one("res.company", default=False)
+    depth = fields.Integer(compute="_compute_would_be_depth")
+
+    def construct(self):
+        for parent_package, links in self.groupby("parent_id"):
+            links.mapped("child_id").write(
+                {"parent_id": parent_package.id if parent_package else False}
+            )
+
+    @api.model
+    def create_top_level_unlinks(self, packages, move_lines=False):
+        # Warm the cache
+        packages.get_move_lines(aux_domain=[("id", "in", move_lines.ids)])
+        link_vals = []
+        # Create unlinks from higher level packages
+        for top_level_pack in packages:
+            if top_level_pack.parent_id:
+                mls = top_level_pack.get_move_lines(aux_domain=[("id", "in", move_lines.ids)])
+                link_vals.append(
+                    {
+                        "parent_id": False,
+                        "child_id": top_level_pack.id,
+                        "move_line_ids": [(6, 0, mls.ids if mls else False)],
+                    }
+                )
+
+        if link_vals:
+            return self.create(link_vals)
+        return self.browse()
+
+    @api.depends("parent_id", "child_id")
+    def _compute_name(self):
+        for link in self.filtered(lambda p: not isinstance(p.id, models.NewId)):
+            link.name = (
+                "Link %s and %s" % (link.parent_id.name, link.child_id.name)
+                if link.parent_id and link.child_id
+                else "Unlink parent of %s" % link.child_id.name
+            )
+
+    @api.constrains("parent_id", "child_id")
+    def constrain_links(self):
+        """Check links against constriants.
+        These need to be checked against associated links as they may interact.
+        """
+        # Get all links that are associated with the same move and/or moveline:
+        move_lines = self.mapped("move_line_ids")  # Direct move lines
+        # Add the move lines of associated moves
+        move_lines |= move_lines.mapped("move_id").mapped("move_line_ids")
+        links = move_lines.mapped("u_result_package_link_ids")
```

Comment:
Would we want to extend this to likes for any move that is "in progress"?

Reply:
Good question. I've only checked against links for other move-lines in the current move but to be thorough we could check against other moves as well. I left it out as I thought it might get very computationally expensive for large systems but it would definitely be 'more correct'. Would you prefer me to add it in or leave it as a TODO?

Reply:
>  Would you prefer me to add it in or leave it as a TODO?

Good question, at least make a note of it here and we can as the `code tsars`